File: tlg/src/main/java/ch/bailu/tlg/Matrix.java
package ch.bailu.tlg;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;

public class Matrix {
    private int width, height;
    private Square[] squares;

    public Matrix(int w, int h) {
        width=w;
        height=h;
        squares = new Square[w*h];
        
        for (int i=0; i<squares.length; i++)
            squares[i] = new Square();
    }
    
    public Matrix(Matrix b) {
        width=b.getWidth();
        height=b.getHeight();
        squares = new Square[b.squares.length];
        
        for (int i=0; i < squares.length; i++)
            squares[i] = new Square(b.squares[i]);
    }
    
    protected int count() {
        return squares.length;
    }
    
    protected Square get(int i) {
        return squares[i];
    }
    
    protected Square get(int x, int y) {
        return get((y * width) + x);
    }

    protected int getWidth()  {return width;}
    protected int getHeight() {return height;}
    protected int getSize() {return Math.max(getWidth(), getHeight());}
    
    public void writeState(BufferedOutputStream output) throws IOException {
        output.write(width);
        output.write(height);
        
        for (int i=0; i< count(); i++) {
            get(i).writeState(output);
        }
    }
    
    public Matrix(BufferedInputStream reader) throws IOException {
        width=reader.read();
        height=reader.read();
        
        squares = new Square[width*height];
        
        for (int i=0; i<count(); i++) {
            squares[i] = new Square(reader);
        }
    }
}

File: tlg/src/main/kotlin/tlg/matrix/Matrix.kt
package tlg.matrix

import tlg.Square
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.IOException

open class Matrix {
    var width: Int
        private set
    var height: Int
        private set
    private var squares = ArrayList<Square>()

    constructor(w: Int, h: Int) {
        width = w
        height = h
        val size = width * height

        for (i in 0 until size) {
            squares.add(Square())
        }
    }

    constructor(b: Matrix) {
        width = b.width
        height = b.height

        b.squares.forEach {
            squares.add(Square(it))
        }
    }


    constructor(reader: BufferedInputStream) {
        width = reader.read()
        height = reader.read()

        for (i in 0 until count()) {
            squares.add(Square(reader))
        }
    }

    fun count(): Int {
        return width * height
    }

    protected operator fun get(i: Int): Square {
        return squares[i]
    }

    operator fun get(x: Int, y: Int): Square {
        return get(y * width + x)
    }

    protected val size: Int
        get() = Math.max(width, height)

    @Throws(IOException::class)
    open fun writeState(out: BufferedOutputStream) {
        out.write(width)
        out.write(height)

        squares.forEach {
            it.writeState(out)
        }
    }
}
